tools/codegen/api/python.py

```diff
@@ -0,0 +1,987 @@
+# They are not always 1-1 mapped, e.g. scattered TensorOptions fields
+# need be packed into a TensorOptions object, which is the argument
+# that the lambda function wrapper takes.
+@dataclass(frozen=True)
+class DispatchLambdaArgumentExprs:
+    # The exprs that provide the binding for lambda arguments, e.g.:
+    #
+    #   'self' -> '_r.tensor(0)'
+    #   'min' -> 'out[0]' / 'min_indices' -> 'out[1]'
+    #   'options' -> 'options'
+    #
+    # It has 1-1 mapping with DispatchLambdaArgument.
+    exprs: Sequence[str]
+
+@dataclass(frozen=True)
+class LocalInitExprs:
+    # The exprs for special local inits, e.g.:
+    #
+    #   'auto out = _r.tensorlist_n<2>(2);'
+    #
+    exprs: Sequence[str]
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                          Helper Functions
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# The original simple_type is derived from the 'type' field in Declaration.yaml,
+# which is generated from the C++ argument type, following some seemingly
+# artificial rules:
+#
+# Concrete C++ types are preferred in most cases, e.g.:
+#   'IntArrayRef' instead of 'int[]'
+#   'int64_t' instead of 'int'
+#
+# Constant/Reference annotation and optional field are handled specially, e.g.:
+#   'ScalarType?' instead of 'c10::optional<ScalarType>'
+#   'Tensor' instead of 'const Tensor &' / 'Tensor &'
+#
+# TODO: This needs to be consistent with python_arg_parser - can we simplify it?
+def _simple_type(cpp_type_str: str) -> str:
+    simple_type = cpp_type_str.replace(' &', '').replace('const ', '')
+    opt_match = re.match(r'c10::optional<(.+)>', simple_type)
+    if opt_match:
+        typename = opt_match.group(1)
+        # HACK: 'Layout?' needs to be hardcoded to 'Layout'!
+        simple_type = f'{typename}?' if typename != 'Layout' else 'Layout'
+    return simple_type
+
+def _cpp_signature(f: NativeFunction, *, method: bool = False) -> cpp.CppSignature:
+    return CppSignatureGroup.from_schema(f.func, method=method).signature
+
+def has_tensor_options(f: NativeFunction) -> bool:
+    return any(filter(lambda a: isinstance(a, TensorOptionsArguments),
+                      cpp.group_arguments(f.func, method=False)))
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                          Python Signature
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+def argument(cpp_arg: CppArgument) -> PythonArgument:
+    a = cpp_arg.argument
+    if not isinstance(a, Argument):
+        # cpp's TensorOptionsArguments is ignored, we will reintroduce the
+        # scattered fields in tensor_options_args.
+        raise RuntimeError(f'unsupported cpp argument: \'{cpp_arg}\'')
+    return PythonArgument(
+        name=a.name,
+        type=a.type,
+        cpp_type_str=cpp_arg.type,
+        default=str(pythonify_default(cpp.default_expr(a.default, a.type)))
```

Comment:
I think I'd probably prefer it if we had a function that directly implemented this conversion from a.default to Python default; Python-ify default was written for the narrow case of getting Declarations.yaml to byte-for-byte serialize correctly, the code here is clever but it is relying on the invariant that defaults produced by cpp.default_expr that are not pythonified are valid defaults for PythonArgParser, which doesn't seem like a very robust invariant.

Reply:
Right, I agree with what you'v suggested on the other comment - I think we should have more direct translation from Type to PythonType (avoid indirect translation via these cpp_type / cpp_default methods).
I haven't got to it yet. I'll work on it next.

Reply:
Added TODO